File: #0138.copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {

        if (!head) {
            return NULL;
        }

        Node* result = new Node(head->val, NULL, NULL);
        Node* n = result;
        Node* cursor = head->next;
        map<Node*,Node*> m;
        Node* work;
        m[head] = result;
        m[NULL] = NULL;
        while (cursor) {
            work = new Node(cursor->val, NULL, NULL);
            n->next = work;
            m[cursor] = work;
            n = n->next;
            cursor = cursor->next;
        }

        n = result;
        cursor = head;
        while (n) { 
            if (!cursor->random) {
                n->random = NULL;
            } else {
                n->random = m[cursor->random];
            }
            n = n->next;
            cursor = cursor->next;
        }
        return result;
    }
};

File: #0138.copy-list-with-random-pointer.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
    const m = new Map();
    let cur = head;
    while (cur) {
        m.set(cur, new Node(cur.val));
        cur = cur.next;
    }

    cur = head;
    while (cur) {
        m.get(cur).next = cur.next ? m.get(cur.next) : null;
        m.get(cur).random = cur.random ? m.get(cur.random) : null;
        cur = cur.next;
    }
    return head ? m.get(head) : null;
};
